File: src/leetCode/java/Solution90.java
package leetCode.java;

import java.util.*;

//Backtracking O(n * 2^n) O(n), 漸進意義上遞迴所需時間複雜度會大於排序所需
class Solution90 {
    public List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);//記得背包問題要先排序
        backtracking(0, nums, new ArrayList<Integer>());
        return res;
    }

    public void backtracking(int i, int[] nums, List<Integer> list){
        if(i == nums.length){
            res.add(new ArrayList<Integer>(list));//O(n)
            return;
        }

        list.add(nums[i]);
        backtracking(i + 1, nums, list);
        list.remove(list.size() - 1);

        //The solution set must not contain duplicate subsets
        while(i + 1 < nums.length && nums[i] == nums[i + 1])
            i++;

        backtracking(i + 1, nums, list);
    }
}
/* 背包問題, 不可以重複 nums = [1,2,2]
				[]
		1				[]
	12		1		2		[]
  122 12  12  1   22  2   2   []
*/

File: src/leetCode/c++/Solution90.cpp
//Backtracking O(n * 2^n) O(n)
class Solution90 {
public:
    vector<vector<int>> res;

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<int> vec;
        sort(nums.begin(), nums.end());//小排到大, 從小開始入列避免重複(此處的重複是指重複使用已用的index), 用來判斷跟前一個數是否一樣
        backtracking(0, vec, nums);
        return res;
    }

    void backtracking(int start, vector<int>& vec, vector<int>& nums) {
        res.push_back(vec);

        for(int i=start; i<nums.size(); ++i){
            if(i > start && nums[i] == nums[i - 1])
                continue;//避免重複, 在遞迴時已經加入了

            vec.push_back(nums[i]);
            backtracking(i + 1, vec, nums);
            vec.pop_back();
        }
    }
};